pandas/tests/indexing/test_indexing.py

```diff
@@ -881,6 +882,14 @@ def test_partial_boolean_frame_indexing(self):
                                 columns=list('ABC'))
         tm.assert_frame_equal(result, expected)
 
+    def test_no_reference_cycle(self):
+        df = pd.DataFrame({'a': [0, 1], 'b': [2, 3]})
+        for name in ('loc', 'iloc', 'ix', 'at', 'iat'):
```

Comment:
Would it make sense to do a `sys.getrefcount(df)` before and after and assert that it doesn't change? Or is this assertion too strong?

Reply:
There is no need for that, the `weakref`-based test below already tests that no leak happens.

Reply:
I was only wondering if the `weakref` can become `None` "by chance" if the garbage collector happens to run exactly between the `del` and the assertion. This is of course highly unlikely, but my understanding was that the `None` result does not directly imply that there wasn't a cyclic reference. But I could be wrong about that.

Reply:
Well, In that case the test would fail most of the time.

Reply:
Why most of the time? It is very unlikely, so it would fail rarely, right?

You can check that this assertion erroneously passes with pandas 0.20.3 where there are definitely cyclic references by forcing an "accidental" GC:

```
In [8]:     def test_no_reference_cycle(happens_to_run_gc):
   ...:         import weakref, gc, sys, pandas as pd
   ...:         df = pd.DataFrame({'a': [0, 1], 'b': [2, 3]})
   ...:         refcount_before = sys.getrefcount(df)
   ...:         for name in ('loc', 'iloc', 'ix', 'at', 'iat'):
   ...:             getattr(df, name)
   ...:         refcount_after = sys.getrefcount(df)
   ...:         print("ref counts {} -> {}".format(refcount_before, refcount_after))
   ...:         wr = weakref.ref(df)
   ...:         del df
   ...:         if happens_to_run_gc:
   ...:             gc.collect()
   ...:         assert wr() is None
   ...:
```

Reply:
You said « the weakref can become None "by chance" ». If that's the only reason the test succeeds, then it would fail most of the time, since the chance of a GC happening between two consecutive opcodes is slim.

I'm not sure what your code snippet is supposed to prove.

Reply:
What I'm trying to say: Checking for `wr() is None` does not guarantee that the operations are ref-cycle-free, which I though was the purpose of the test. That's why I would have preferred `refcount_before == refcount_after` but due to the low probability of an accidental success it doesn't really matter.